[RPC] fundrawtransaction

jonasschnelli opened the pull request:
Basic implementation of a new rpc call `fundrawtransaction`.
This call will fill a rawtransaction containing only vouts with unspent coins from the wallet.

Currently there is no way to return the `CReserveKey` to the keypool.

jonasschnelli committed:
[RPC] fundrawtransaction basics

jonasschnelli committed:
[RPC] add support for existing vins for `fundrawtransaction`

jonasschnelli committed:
[RPC] add simple unittest for `fundrawtransaction` (increase test coverage)

jonasschnelli committed:
[RPC] fundrawtransaction overhaul

- fix typo
- fix RPCTypeCheck for fundrawtransaction
- some comments
- merged with subtractFeeFromAmout patch (non trivial)

laanwj commented:
needs rebase

jonasschnelli commented:
rebased.

jonasschnelli commented:
should solve #3794

jonasschnelli commented:
binaries to test: https://builds.jonasschnelli.ch/pulls/5503/

jonasschnelli commented:
Rebased (hard rebase after "Subtract fee from amount [#5831]"), added some code comments.

jgarzik commented on src/rpcserver.cpp:
```diff
@@ -321,7 +321,9 @@ static const CRPCCommand vRPCCommands[] =
     { "rawtransactions",    "signrawtransaction",     &signrawtransaction,     false,     false }, /* uses wallet if enabled */
-
+#ifdef ENABLE_WALLET
+    { "rawtransactions",    "fundrawtransaction",     &fundrawtransaction,     false,     false },
```
Bug:  should be 'true' to indicate wallet requirement.  Otherwise, the 'pwalletMain' reference will oops when NULL (wallet support built, but disabled at runtime).

jonasschnelli replied:
Thanks for catching this one! Fixed and pushed.

jgarzik commented:
ut ACK + plan to test this this week.

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -1823,6 +1893,14 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend,
                     return false;
                 }
+                
+                //remove signature if we used the signing only for the fee calculation
```
Why sign at all if it's only for the fee calculation? I ask because we don't want to unnecessarily require the signing keys.

jonasschnelli replied:
We still need to remove the signature, but now, during fundrawtransaction, the signature is dummy/void.

laanwj replied:
Yes, removing a dummy sigature is fine.

jonasschnelli commented:
Currently i'm working on this to allow fundrawtransaction to calculate serialized transaction length for getting the right fee and priority without actually signing the tx. Tests won't make sense at the moment.

jonasschnelli committed:
[RPC] fundrawtransaction dummy signing / ser. length calculation

- serialize transaction size calculation (aka dummy-signing) without the need of signing the transaction (wallet can be locked to use fundrawtransaction)
- rename VINS to vInputs

jonasschnelli commented:
Updated!
`fundrawtransaction` now works with locked wallets. Fee / priority calculation now goes over a new dummy-signing way in `sign.cpp`.
RPC tests are now even more extensive and includes multisig test, locked wallet tests as well as comparison of calculated fee and correct fee tests.

The only question is, if it would be worth to add a optional `correctFee=1|0` argument for the `fundrawtransaction` rpc call. With settings this flag, users could enforce correct signing of the tx (would require unlocked wallet) which would end up in getting a more precise/correct fee.

jonasschnelli committed:
[moveonly] move fundrawtransaction() rpc call to rpcwallet.cpp

fundrawtransaction requires a wallet and therefore it belongs to rpcwallet.cpp

jonasschnelli committed:
add wallet-is-present check for fundrawtransaction

rpc dispatch tables reqWallet has been removed.

jonasschnelli commented:
rebased.

Moved RPC function "fundrawtransaction()" from `rpcrawtransaction.cpp` to `rpcwallet.cpp` (fundrawtransaction is a wallet function). Added wallet-is-presence check because we recently removed "reqWallet" check over the RPC dispatch table.

jonasschnelli committed:
[QA] adapt generate instead of setgenerate for fundrawtransaction tests

TheBlueMatt committed:
Fix whitespace errors

# Conflicts:
#	src/rpcserver.cpp

jonasschnelli commented:
rebased.

TheBlueMatt committed:
Tweak fundrawtransaction help text and return the fee as a float

TheBlueMatt committed:
Test inputs are not modified in fundrawtransaction

TheBlueMatt committed:
Also return changepos from fundrawtransaction

# Conflicts:
#	src/wallet/rpcwallet.cpp

TheBlueMatt commented:
I needed to use this for something and rewrote/tweaked a few chunks. See changes at https://github.com/TheBlueMatt/bitcoin/tree/frt

jonasschnelli commented:
Pulled in some commits from @TheBlueMatt. Only pulled in clear beneficial and easy reviewable commits. Other things (CCoinControl usage, watch-only support) may be better handled in a 2nd pull request to avoid large changeset.

TheBlueMatt commented:
@jonasschnelli Actually, the changes I made were mostly to limit the size of *this* pull request. Some of them look like it would become a big pull request, but once squashed, the total size is actually smaller than the original :)

TheBlueMatt commented:
Also, not sure when you pulled in, but I force pushe'd a few times, you may want to check that the ones you merged are the latest versions (I'm done now, I think). I also merged in a tweaked-up version of fundrawtransaction there.

jonasschnelli commented:
@TheBlueMatt Thank you for your tweaks! Will try to go through the optimizing commits but need a clear head for that. Will do it soon.

TheBlueMatt commented:
I added two more commits to that tree and made a nice rebased version at https://github.com/TheBlueMatt/bitcoin/commits/frt2

TheBlueMatt commented:
Rebased frt2. @jonasschnelli do you want to review my changes and merge them here, or should I just open a new pull request? I'd like to keep this moving.

jonasschnelli commented:
closing in favor of #6088
